packages/scripts/src/split-file/cursor.js

```diff
@@ -0,0 +1,234 @@
+import { readFileSync } from 'fs';
+
+/**
+ * @typedef {'none' | 'in-import' | 'in-type' | 'in-codeblock' | 'end'} CursorState
+ * @typedef {'empty' | 'filling' | 'contains-import' | 'contains-type' | 'contains-codeblock'} FragmentState
+ */
+
+const IMPORT_START_INDICATOR = Buffer.from('import');
+const TYPE_START_INDICATOR = Buffer.from('/**');
+const BLOCK_START_INDICATORS = ['const', 'export', 'async', 'function'].map((str) =>
+	Buffer.from(str)
+);
+
+const IMPORT_END_INDICATOR = Buffer.from(';');
+const TYPE_END_INDICATOR = Buffer.from('*/');
+
+export class Cursor {
```

Comment:
This guy goes through a file character by character and feeds back what kind of code it's reading (import, type declaration or code block). It fills an internal buffer with the element it's currently reading hen flags when that buffer contains something worth reading out. 

Currently it contains all the rules for distinguishing between different types of block, which makes it look quite gungey. I'd like to extract those but it didn't feel necessary now.